Netatmo integration

Terdious opened the pull request:
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] Did you test this pull request in real life? With real devices? If this development is a big feature or a new service, we recommend that you provide a Docker image to the community ([french forum](https://community.gladysassistant.com/)/[english forum](https://en-community.gladysassistant.com/)) for testing before merging.
- [x] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Add a new Netatmo integration
OAuth2 authentication management

Terdious committed:
Netatmo integration frontend & constants

Terdious committed:
Server with events and access_token connection

Terdious committed:
Refactor front connection access_token

Terdious committed:
front refactoring, improvement of status feedback and i18n update

Terdious committed:
refactor server + eslint + first tests

Terdious committed:
Front prettier

Terdious committed:
Tests connect.js

Terdious committed:
Continu Tests : index - mock - connect - disconnect - init - controller

Terdious committed:
Continuation of tests and related modifications

Terdious committed:
fix error tests with 'sinon.restore' and 'spy.restore'

Terdious committed:
fix eslint + prettier + tests

Terdious committed:
fix prettier front

Terdious committed:
cleaning and adding discovery/configuration/viewing/control of devices

Terdious committed:
Merge branch 'master' into netatmo-integration

Terdious committed:
following first returns fix remarks and errors

Terdious committed:
Merge branch 'master' into netatmo-integration

Terdious committed:
front change following the remark of the 1st test. Simplifying the connection.

Terdious committed:
server changes following the remarks of the 1st test. Modification of the features released for v1.

Terdious committed:
forgetting utils/constants

Terdious committed:
last add front and server

Terdious committed:
Merge branch 'master' into netatmo-integration

Terdious committed:
refactor: modify loadThermostat code and i18n

Terdious committed:
refactor: clean-up of the global front

Terdious committed:
fix: prettier + eslint (front)

Terdious committed:
refactor (front): deletion of unused translations, addition of unsupported devices, modification of reachable device method, deletion of features

Terdious committed:
refactor (server): adding automation of scopes, adding unsupported devices, removing features

Terdious committed:
refactor: forgotten deletions i18n

Terdious committed:
test: modifications following server code changed

Terdious committed:
test: continuation of test corrections

Terdious committed:
test: continuation of tests and server code fixes following completion of tests

Terdious committed:
fix: tests

Terdious committed:
fix: tests

Terdious committed:
Merge branch 'master' into netatmo-integration

Terdious committed:
tests: continuation of test writing

Terdious committed:
fix: tests prettier and increase coverage

Terdious committed:
test: improvement of tests with interval and taking into account tests of new constants

Terdious committed:
fix: test on errors Timeout exceeded other integration

Terdious committed:
Merge branch 'master' into netatmo-integration

Terdious committed:
refactor: added German translate

Terdious committed:
refactor: modifying links to sites in the correct language

Terdious committed:
Merge branch 'master' into netatmo-integration

Terdious committed:
Merge branch 'master' into netatmo-integration

Pierre-Gilles commented on front/src/assets/integrations/devices/netatmo/netatmo-NAPlug.jpg:
This image is too big :) You need to resize+compress it. This kind of image could be easily 50ko

Pierre-Gilles commented on front/src/assets/integrations/devices/netatmo/netatmo-NATherm1.jpg:
This image is too big :) You need to resize+compress it. This kind of image could be easily 50ko

Pierre-Gilles commented on front/src/routes/integration/all/netatmo/device-page/index.js:
```diff
@@ -0,0 +1,89 @@
+  };
+
+  componentDidMount() {
+    // this.loadProps();
```
Unused code should be removed

Pierre-Gilles commented on front/src/routes/integration/all/netatmo/setup-page/style.css:
```diff
@@ -0,0 +1,25 @@
+.btnTextLineSpacing {
+  line-height: 1.5;
+  /* La valeur peut être ajustée selon les besoins, 1 est le défaut */
```
french comment, should be in english

Pierre-Gilles commented on server/services/netatmo/api/netatmo.controller.js:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  /**
+   * @api {post} /api/v1/service/netatmo/saveConfiguration Save Netatmo Configuration.
```
our API is snake_case + following the REST principle ( https://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api ) 

In that case, it should be: 

```
POST /api/v1/service/netatmo/configuration
```

Pierre-Gilles commented on server/services/netatmo/api/netatmo.controller.js:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  /**
+   * @api {post} /api/v1/service/netatmo/saveStatus save Netatmo connection status
```
Same here: 

```
POST /api/v1/service/netatmo/status
```

Pierre-Gilles commented on server/services/netatmo/api/netatmo.controller.js:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  /**
+   * @api {post} /api/v1/service/netatmo/retrieveTokens Retrieve access Tokens netatmo with code of return
```
Same here, if you do a get, you should do: 

```
GET /api/v1/service/netatmo/token
```

Terdious replied:
It's well POST here but ok for "POST /api/v1/service/netatmo/token"

Pierre-Gilles commented on server/services/netatmo/api/netatmo.controller.js:
```diff
@@ -0,0 +1,152 @@
+    await netatmoHandler.getConfiguration(netatmoHandler);
+    const result = await netatmoHandler.retrieveTokens(netatmoHandler, req.body);
+    res.json({
+      result,
```
why wrap the result in a "result" attribute and not just return the object? Avoid un-necessary wrappers :)

Pierre-Gilles commented on server/services/netatmo/api/netatmo.controller.js:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  /**
+   * @api {get} /api/v1/service/netatmo/refreshDiscover Retrieve netatmo devices from API.
```
Same here, avoid `refreshDiscover`.

Maybe: 

```
GET /api/v1/service/netatmo/device
```

?

Pierre-Gilles commented on server/services/netatmo/lib/device/examples.js:
```diff
@@ -0,0 +1,295 @@
+/* console.log in DiscoveredDevice
```
Old file?

Pierre-Gilles commented on server/services/netatmo/lib/index.js:
```diff
@@ -0,0 +1,85 @@
+
+/**
+ *
+ * @description Poll refreshing Token values of an Netatmo device.
```
Wrong comment?

Pierre-Gilles commented on server/services/netatmo/lib/index.js:
```diff
@@ -0,0 +1,85 @@
+  });
+
+  return scopesConfig;
+}
```
could this function be moved to another file?

Terdious replied:
Ok

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.connect.js:
```diff
@@ -0,0 +1,102 @@
+  };
+  try {
+    const response = await axios({
+      url: `${API.TOKEN}`,
```
Why wrap this variable in a string with no other content around? Is it to convert it to String?

Terdious replied:
My bad

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.discoverDevices.js:
```diff
@@ -0,0 +1,50 @@
+ * @example
+ * await discoverDevices(netatmoHandler);
+ */
+async function discoverDevices(netatmoHandler) {
```
why do you pass the current object as param of the function? You are inside an object so you have access to `this`

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.getConfiguration.js:
```diff
@@ -0,0 +1,34 @@
+ * @example
+ * await getConfiguration();
+ */
+async function getConfiguration(netatmoHandler) {
```
Same

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.getTokens.js:
```diff
@@ -0,0 +1,78 @@
+ * @example
+ * await netatmo.getAccessToken(netatmoHandler);
+ */
+async function getAccessToken(netatmoHandler) {
```
same

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.getTokens.js:
```diff
@@ -0,0 +1,78 @@
+ * @returns {Promise} Netatmo refresh token.
+ * @example
+ * await netatmo.getRefreshToken(netatmoHandler);
+ */
```
Can you separate this file into 2 files? 

1 function = 1 file

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.saveValue.js:
```diff
@@ -0,0 +1,86 @@
+ * @param {object} deviceGladys - Of nothing.
+ * @param {object} deviceNetatmo - Of nothing.
+ * @param {string} externalId - Of nothing.
+ * @returns {Promise} Promise of nothing.
```
Bad comment?

Pierre-Gilles commented on front/src/routes/integration/all/netatmo/NetatmoDeviceBox.jsx:
```diff
@@ -0,0 +1,407 @@
+    const sidDevice = device.external_id.replace('netatmo:', '') || (device.deviceNetatmo && device.deviceNetatmo.id);
+    const saveButtonCondition =
+      (saveButton && !alreadyCreatedButton) || (saveButton && !this.state.isSaving && alreadyCreatedButton);
+    const modelImage = `../../../../assets/integrations/devices/netatmo/netatmo-${device.model}.jpg`;
```
Be careful with relative path, in other parts of the code we use absolute path

Terdious replied:
Ok, corrected

Pierre-Gilles reviewed (changes requested):
Thanks for this PR 🙏 

I haven't entirely tested the integration, I just did a first review with what I could test on the UI side + a first code review.

**On the functional side**

When I first open the integration, here is what I see:

![Screenshot 2024-01-11 at 09 54 11](https://github.com/GladysAssistant/Gladys/assets/7365207/2efc8228-8fb8-4fcc-bcb6-1a9270def362)

There is an error message immediately (I didn't do anything), which is a bit weird for the user.

Link to the documentation is missing category: 
![Screenshot 2024-01-11 at 09 54 30](https://github.com/GladysAssistant/Gladys/assets/7365207/a3a997aa-472f-4390-823b-3d07bd67039f)

Buttons a bit weird:

![Screenshot 2024-01-11 at 09 55 08](https://github.com/GladysAssistant/Gladys/assets/7365207/9d6b9657-6266-4ea1-a68e-47725821f83b)

**On the code side**

The code is so much different from when I first had a look 😅

- There is a major issue with the way the `NetatmoHandler` works, it seems you are passing the object (itself) to every single function. It's an anti-pattern, you have access to `this` everywhere. 
- Many files of this class contains several functions. Try to keep 1 file = 1 function when you are building functions of a class (Talking about NetatmoHandler)
- Some function comments are not explaining what a function does, making the code hard to grasp.
- See all my comments

Please take some time to review the PR entirely and improve those points before I can do a second review. 

This is already a massive PR (+7000 lines of code) so each review is expensive.

Terdious committed:
Merge branch 'master' into netatmo-integration

Terdious committed:
refactor: first fixes server code

Terdious committed:
refactor: first fixes front

Terdious committed:
refactor-fix: front corrections

Terdious committed:
refactor-fix: server corrections + tests

Terdious committed:
Merge branch 'master' into netatmo-integration

Terdious committed:
refactor: add front message update values

Terdious committed:
refactor: Update Netatmo setup page styles and text display - using this.connected

Terdious committed:
refactor: Update Netatmo device discovery messages

Terdious committed:
refactor: Add StateConnection component to DeviceTab

Terdious committed:
refactor+fix: Fix console.log statements and optimize devicePage and discovery

Terdious committed:
fix: Fix syntax error in DeviceTab.jsx

Terdious committed:
refactor+fix: Remove console.log statements and add/modify alert messages

Terdious committed:
refactor: Update Netatmo device and discover page texts + deleted texts i18n not use

Terdious committed:
refactor: Netatmo device features update method

Terdious committed:
refactor: Update and Tests NATherm1 and NAPlug features

Terdious committed:
refactor: NetatmoDeviceBox component and style.css

This commit refactors the NetatmoDeviceBox component by replacing a div class with "page-options d-flex" and removes the .firmware-revision style from the style.css file.

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.loadDevices.js:
```diff
@@ -0,0 +1,43 @@
+        }
+        return undefined;
+      });
+      const results = await Promise.all(devicesPromises);
```
You are doing a Promise.all on all devices ? It seems that each loadDeviceDetails call the API, so if the user has 50 Netatmo devices, it'll start 50 API calls in parallel, it's way to heavy!

Consider using bluebird Promise.map instead so you can specify a max concurrency :) 

```
const results = await Promise.map(homes, async(home) => 
... 
}, {concurrency: 2} );
```

Terdious replied:
I'm understand, There is normally not too much risk in this regard because the requests are unique for each house. It's very unlikely that a user will have more than 5 houses so I didn't do too much research. But this way it's much cleaner. THANKS.

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.loadDeviceDetails.js:
```diff
@@ -0,0 +1,139 @@
+      listDevices = modulesHomestatus
+        ? (
+            await Promise.all(
+              modulesHomestatus.map(async (module) => {
```
I don't find any async code in this function? If it's sync, use just .map instead of Promise.all

Terdious replied:
Indeed !!

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.updateValues.js:
```diff
@@ -0,0 +1,37 @@
+  }
+  switch (deviceNetatmo.type) {
+    case SUPPORTED_MODULE_TYPE.PLUG: {
+      await updateNAPlug.bind(this)(deviceGladys, deviceNetatmo, externalId);
```
Weird pattern, why bind "thi"s instead of just having the function in the netatmo object if you need "this"? 

If you don't need this, just build it as a helper function

Terdious replied:
Indeed, it simplifies... I assumed it would clutter the netatmo object... Sorry.

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.updateValues.js:
```diff
@@ -0,0 +1,37 @@
+      break;
+    }
+    case SUPPORTED_MODULE_TYPE.THERMOSTAT: {
+      await updateNATherm1.bind(this)(deviceGladys, deviceNetatmo, externalId);
```
Same

Pierre-Gilles commented on server/services/netatmo/lib/netatmo.pollRefreshingValues.js:
```diff
@@ -0,0 +1,45 @@
+    });
+    logger.error('Unable to load Netatmo devices', e);
+  }
+  devicesNetatmo.map(async (device) => {
```
You cannot do async work in a .map (sync code)

You need to use bluebird Promise.map instead. Don't forget to set a concurrency

Terdious replied:
How to evaluate the competition? Is 10 really suitable? I admit I don't really know. In any case I currently have no lag

Pierre-Gilles replied:
I would recommended maximum of 2 for something background like this.

Pierre-Gilles reviewed (changes requested):
Thanks for the fixes! 🙏

It's way better now :) 

I have a few feedbacks on the server side but nothing serious.

On the UI, it looks good to me now! Can you update your documentation PR with new screenshots?

Once everything is fixed, it'll good for merge I think !! 🚀🚀🚀

Terdious committed:
fix: Following the review add bluebird dependency and update device promises functions

Terdious committed:
Merge branch 'master' into netatmo-integration

Terdious committed:
refactor: Update concurrency in refreshNetatmoValues function

Terdious committed:
refactor: coverage Netatmo devices update functions

Terdious committed:
fake: push with error to relaunch codecov

Terdious committed:
fake: push without error to relaunch codecov

Pierre-Gilles reviewed (approved):
Good for me for this first PR, thanks for this great work 👏👏👏
